src/lib.rs

```diff
@@ -1,9 +1,3221 @@
+                literal_uppercase,
+                // flags,
+                minimum_required,
+            }
+        }
+    }
+
+    impl MatchNotRange {
+        pub(crate) fn new(
+            next : Box<dyn Match>,
+            characters : String,
+            _flags : i64,
+        ) -> Self {
+            // NOTE: this is a not-currently-implemented feature
+            let minimum_required = usize::MAX;
+
+            Self {
+                next,
+                characters,
+                // flags,
+                minimum_required,
+            }
+        }
+    }
+
+    impl MatchRange {
+        pub(crate) fn new(
+            next : Box<dyn Match>,
+            characters : String,
+            _flags : i64,
+        ) -> Self {
+            // NOTE: this is a not-currently-implemented feature
+            let minimum_required = usize::MAX;
+
+            Self {
+                next,
+                characters,
+                // flags,
+                minimum_required,
+            }
+        }
+    }
+
+    impl MatchWild1 {
+        pub(crate) fn new(next : Box<dyn Match>) -> Self {
+            // NOTE: this is a not-currently-implemented feature
+            let minimum_required = usize::MAX;
+
+            Self {
+                next,
+                minimum_required,
+            }
+        }
+    }
+
+    impl MatchWildN {
+        pub(crate) fn new(next : Box<dyn Match>) -> Self {
+            // NOTE: this is a not-currently-implemented feature
+            let minimum_required = usize::MAX;
+
+            Self {
+                next,
+                minimum_required,
+            }
+        }
+    }
+
+    // Trait implementations
+
+    impl Match for MatchEnd {
+        fn matches(
+            &self,
+            slice : &str,
+        ) -> bool {
+            slice.is_empty()
+        }
+    }
+
+    impl Match for MatchLiteral {
+        fn matches(
+            &self,
+            slice : &str,
+        ) -> bool {
+            let slice_starts_with_literal = slice.starts_with(&self.literal)
+                || match &self.literal_uppercase {
+                    Some(literal_uppercase) => {
+                        if slice.len() >= literal_uppercase.len() {
+                            slice.to_uppercase().starts_with(literal_uppercase)
+                        } else {
+                            false
+                        }
+                    },
+                    None => false,
+                };
+
+            if !slice_starts_with_literal {
+                return false;
+            }
+
+            let next = self.next.as_ref();
+
+            next.matches(&slice[self.literal.len()..])
+        }
+    }
+
+    impl Match for MatchNotRange {
+        fn matches(
+            &self,
+            slice : &str,
+        ) -> bool {
+            if slice.is_empty() {
+                return false;
+            }
+
+            let c0 = slice.chars().next().unwrap();
```

Comment:
It's possible to avoid having an `unwrap()`.
```suggestion
            let c0 = match slice.chars().next() {
                Some(c0) => c0,
                None => return false,
            };
```